hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestPartialRead.java

```diff
@@ -0,0 +1,260 @@
+    try {
+      stream.write(b);
+    } finally {
+      stream.close();
+    }
+    return b;
+  }
+
+  @Test
+  public void testRecoverPartialRead() throws Exception {
+    int fileSize = 4 * ONE_MB;
+    Path testPath = path(TEST_PATH);
+    byte[] originalFile = setup(testPath, fileSize);
+
+    final AzureBlobFileSystem fs = getFileSystem();
+    MockAbfsClient abfsClient = new MockAbfsClient(fs.getAbfsClient());
+
+    ActualServerReadByte actualServerReadByte = new ActualServerReadByte(
+        fileSize, originalFile);
+    MockHttpOperationTestIntercept mockHttpOperationTestIntercept
+        = new MockHttpOperationTestIntercept() {
+      private int callCount = 0;
+
+      @Override
+      public MockHttpOperationTestInterceptResult intercept(final MockHttpOperation mockHttpOperation,
+          final byte[] buffer,
+          final int offset,
+          final int length) throws IOException {
+        /*
+         * 1. Check if server can handle the request parameters.
+         * 2. return 1MB data to test-client.
+         */
+        callActualServerAndAssertBehaviour(mockHttpOperation, buffer, offset,
+            length, actualServerReadByte, ONE_MB);
+
+        MockHttpOperationTestInterceptResult
+            mockHttpOperationTestInterceptResult
+            = new MockHttpOperationTestInterceptResult();
+        mockHttpOperationTestInterceptResult.setStatus(HTTP_PARTIAL);
+        mockHttpOperationTestInterceptResult.setBytesRead(ONE_MB);
+        callCount++;
+        return mockHttpOperationTestInterceptResult;
+      }
+
+      public int getCallCount() {
+        return callCount;
+      }
+    };
+    abfsClient.setMockHttpOperationTestIntercept(
+        mockHttpOperationTestIntercept);
+    fs.getAbfsStore().setClient(abfsClient);
+
+    AbfsClientThrottlingIntercept intercept
+        = AbfsClientThrottlingInterceptTestUtil.get();
+    MockAbfsClientThrottlingAnalyzer readAnalyzer
+        = new MockAbfsClientThrottlingAnalyzer("read");
+    MockAbfsClientThrottlingAnalyzer analyzerToBeAsserted
+        = (MockAbfsClientThrottlingAnalyzer) AbfsClientThrottlingInterceptTestUtil.setReadAnalyzer(
+        intercept, readAnalyzer);
+
+    FSDataInputStream inputStream = fs.open(testPath);
+    byte[] buffer = new byte[fileSize];
+    inputStream.read(0, buffer, 0, fileSize);
+
+    Assert.assertEquals(4, mockHttpOperationTestIntercept.getCallCount());
+    Assert.assertEquals(4,
+        analyzerToBeAsserted.getFailedInstances().intValue());
+  }
+
+  private void callActualServerAndAssertBehaviour(final MockHttpOperation mockHttpOperation,
+      final byte[] buffer,
+      final int offset,
+      final int length,
+      final ActualServerReadByte actualServerReadByte,
+      final int byteLenMockServerReturn) throws IOException {
+    LOG.info("length: " + length + "; offset: " + offset);
+    mockHttpOperation.processResponseSuperCall(buffer, offset, length);
+    Assert.assertTrue(
```

Comment:
Can use static Assert in place of junit asserts.

Reply:
Refactored.